### PR TITLE
Avoid downloading images if already in the database

### PR DESCRIPTION
**Pull request by harry-73**
and adding an option 'alwaysDownload' to able or not (default) this feature.

### CODE REVIEW

File: src/TileManager.js [hunk 1]
```diff
@@ -78,13 +78,25 @@ export async function getStorageInfo(urlTemplate) {
...
  * @return {Promise<blob>}
  */
-export async function downloadTile(tileUrl) {
-  return fetch(tileUrl).then((response) => {
-    if (!response.ok) {
-      throw new Error(`Request failed with status ${response.statusText}`);
-    }
-    return response.blob();
-  });
+export async function downloadTile(tileUrl,tileKey, alwaysDownload) {
```
Comment (allartk) [inline src/TileManager.js:81]: Thanks for your PR! Please separate responsibility here, checking if the storage has a tile and downloading are two very different things.

Comment (allartk) [reply src/TileManager.js:81]: Also please adapt coding standards of this repo

**Comment by harry-73** [comment]
Please don't review it. I will pull corrections ...

File: src/ControlSaveTiles.js [hunk 2]
```diff
@@ -225,14 +227,37 @@ const ControlSaveTiles = L.Control.extend(
...
-        self._saveTile(tile, blob);
-        self._baseLayer.fire('loadtileend', self.status);
-        if (self.status.lengthLoaded === self.status.lengthToBeSaved) {
-          self._baseLayer.fire('loadend', self.status);
-        }
-      });
+
+      if (this.options.alwaysDownload) {
+        await downloadTile(tile.url).then((blob) => {
+          self.status.lengthLoaded += 1;
```
Comment (allartk) [inline src/ControlSaveTiles.js:233]: Thanks for your improvements to the code! I see 3 identical blocks here, can you make it one (eg add some method)

File: src/TileManager.js [hunk 3]
```diff
@@ -78,13 +78,14 @@ export async function getStorageInfo(urlTemplate) {
...
  * @return {Promise<blob>}
  */
-export async function downloadTile(tileUrl) {
-  return fetch(tileUrl).then((response) => {
+ export async function downloadTile(tileUrl) {
+  return fetch(tileUrl).then(response => {
     if (!response.ok) {
-      throw new Error(`Request failed with status ${response.statusText}`);
-    }
+      throw new Error("Request failed with status ".concat(response.statusText));
```
Comment (allartk) [inline src/TileManager.js:84]: What is the reason for this change? Also i am not sure, a concat function exists (it exists as method on various objects)

**Review by allartk — CHANGES_REQUESTED** [review]

File: src/TileManager.js (hunk 3 above)
Comment (harry-73) [reply src/TileManager.js:84]: There is no reason. A mistaking on my side

**Review by allartk — APPROVED** [review]
